### ISSUE TITLE
Support invoking objdump on binary files for 'MinGW GCC' and 'Cross GCC' toolchains

### ISSUE THREAD

**Issue by jld01** [eclipse-cdt/cdt#361]
When building CDT projects using a native _Linux GCC_ or _Cygwin GCC_ toolchain, the user can double click on the resulting binary file to invoke `objdump` and present the resulting output in a read-only editor part. It would be beneficial to provide equivalent behaviour when using _MinGW GCC_ toolchains and _Cross GCC_ toolchains.

For  _Linux GCC_, this capability is implemented via the `DefaultGnuToolFactory` and CDT expects to find `objdump` on the global `PATH`.

A few issues:
- _Cross GCC_ toolchains use the `org.eclipse.cdt.core.ELF` binary parser rather than the `org.eclipse.cdt.core.GNU_ELF` binary parser at present. Easy to fix.
- We could create an equivalent `org.eclipse.cdt.core.GNU_PE64` binary parser for _MinGW GCC_ toolchains. Easy to fix.
- _Cross GCC_ toolchains will typically provide a build of `objdump` with a prefix (eg `arm-none-eabi-objdump`) but the `IGnuToolFactory` implementation currently has no context from which to determine that prefix.
- It is much less common for GNU tools to be located on the global `PATH` on Windows hosts or when using a cross compiler. We need a mechanism to lookup the project containing the binary file and from there lookup the per-project `PATH`.

**Comment by jantje** [comment]
my 2 cent

>It is much less common for GNU tools to be located on the global PATH on Windows hosts or when using a cross compiler. We need a mechanism to lookup the project containing the binary file and from there lookup the per-project PATH.

You are probably aware but to be sure: The current MBS allows to extend the path environment variable in the settings.
![afbeelding](https://user-images.githubusercontent.com/1175841/230973074-88aeba00-08dd-4b3b-a8d8-e23a146d6de7.png)

In Arduino world cross GCC, MinGW, MVC are called **tool providers**. These tool providers have a "tool path" property.
All commands are executed with FQN (in MBS talk ${provider tool path including /}${COMMAND}).
An environment variable provider can provide a value for the variable "provider tool path including /".

I have had no luck with the path extension used by MBS. I like the FQN approach.

**Comment by jld01** [comment]
Thank you for your input, @jantje.

Ideally we should accommodate both MBS and non-MBS projects, both with or without a tool prefix. The `IGnuToolFactory` implementation could look up a provider of `PATH` and tool prefix information for a certain binary file path. There would need to be a more-specialized provider for the case of an MBS _Cross GCC_ toolchain project that understood the MBS `IOption` providing the prefix string. It seems that a common provider could be used for most other CDT projects (both MBS and non-MBS).

**Comment by Torbjorn-Svensson** [comment]
>     * It is much less common for GNU tools to be located on the global `PATH` on Windows hosts or when using a cross compiler. We need a mechanism to lookup the project containing the binary file and from there lookup the per-project `PATH`.

I don't think "per-project" will do here as there might be more than one toolchain used within one CDT project.
In embedded, it's not uncommon to have one build configuration per target permutation. It's also fairly common that you have some unit tests that might be executed on the developer system rather than on the target (PC toolchain instead of an embedded one). In this scenario, at least 2 different objdump instances needs to be pointed out for the single CDT project.

**Comment by jld01** [comment]
> I don't think "per-project" will do here...

Totally agree, @Torbjorn-Svensson. I have found a way to lookup an `ICConfigurationDescription` object and hence the per-configuration environment from `DefaultGnuToolFactory`. From here, I should be able to adapt to something with knowlege of the MBS cross toolchain that can return a toolchain prefix for a specified build configuration.

**Comment by Torbjorn-Svensson** [comment]
> > I don't think "per-project" will do here...
> 
> Totally agree, @Torbjorn-Svensson. I have found a way to lookup an `ICConfigurationDescription` object and hence the per-configuration environment from `DefaultGnuToolFactory`. From here, I should be able to adapt to something with knowledge of the MBS cross toolchain that can return a toolchain prefix for a specified build configuration.

Is there/Will there be any easy way for downstream to override how the path and prefix should be collected for a custom toolchain definition?
It would be a bit sad to put in all this effort to support a few more toolchain definitions without having a generic way to allow downstream to hook in their custom logic to determine the right value for the 2 properties in their use-case.

**Comment by jld01** [comment]
> Is there/Will there be any easy way for downstream to override how the path and prefix should be collected for a custom toolchain definition?

Yes, I'm working on an optional CDT build variable to provide the GNU tool prefix so that:

- Any MBS toolchain can provide the CDT build variable via an `IConfigurationBuildMacroSupplier`
- MBS toolchains derived from the CDT _Cross GCC_ toolchain can inherit the _Cross GCC_ toolchain's implementation
- The end user can provide their own CDT build variable for legacy MBS toolchains and non-MBS cases
- The build _variable_ does not automatically polute the build _environment_ but the user has the flexibility to define an environment variable such as `CROSS_COMPILE` based on the build variable if required

**Comment by jld01** [comment]
> Is there/Will there be any easy way for downstream to override how the path and prefix should be collected for a custom toolchain definition?

@Torbjorn-Svensson does #368 address your concerns? Extenders at the `org.eclipse.cdt.managedbuilder.core.buildDefinitions` extension point will be able to use an `IConfigurationBuildMacroSupplier` to provide an `IBuildMacro` named `gnu_tool_prefix`. For many such extensions, a String literal value for the macro (eg `arm-none-eabi-`) will be sufficient.

**Comment by Torbjorn-Svensson** [comment]
> @Torbjorn-Svensson does #368 address your concerns? Extenders at the `org.eclipse.cdt.managedbuilder.core.buildDefinitions` extension point will be able to use an `IConfigurationBuildMacroSupplier` to provide an `IBuildMacro` named `gnu_tool_prefix`. For many such extensions, a String literal value for the macro (eg `arm-none-eabi-`) will be sufficient.

I took a look at the PR and overall, it looks good. There were a few small questions/suggestions. Let's handle them in the PR and I'm sure we will get a solution that will be extendable for downstream vendors.

**Comment by jld01** [comment]
The initial objectives described in this issue are now met. I will work on a _New & Noteworthy_ entry.

There is an outstanding item for discussion. Namely, a mechanism for updating the binary parser selection for existing _MinGW GCC_ and _Cross GCC_ projects:

- `PE64` -> `GNU_PE64`
- `ELF` -> `GNU_ELF`

**Comment by baoruiqi** [comment]
@jld01 Thanks for your effort, this is a very useful feature! Based on this, I have a question.

I downloaded the latest Embedded IDE from Eclipse's website, created a C managed project and selected the riscv-cross toolchain. In the project properties, I set the local toolchain path, and the binary file of the risc-v architecture can be successfully generated after the build.

However, when I double-click the elf file, the editor view only shows the header information, and there is no disassembly content.

![image](https://github.com/eclipse-cdt/cdt/assets/128341204/a3b026b4-43af-489a-9afe-ce68c535cc34)

Usually the disassembly is obtained through the objdump tool related to the toolchain, passing the -d parameter. However, the binary parser tab in the project configuration interface provided by CDT does not provide relevant parameter configuration entries.

If I want to double-click the binary file and display the disassembly, do I need to modify the cdt?

Below is my IDE info:
```
Eclipse IDE for Embedded C/C++ Developers (includes Incubating components)
Version: 2023-06 M2 (4.28.0 M2)
Build id: 20230504-1428
OS: Linux, v.5.19.0-40-generic, x86_64 / gtk 3.24.33, WebKit 2.38.6
Java vendor: Eclipse Adoptium
Java runtime version: 17.0.7+7
Java version: 17.0.7
```

**Comment by jld01** [comment]
@baoruiqi, the `architecture: UNKNOWN!` output suggests that you are invoking a native `objdump` that exists somewhere on your PATH.

When using the _Embedded CDT_ toolchain descriptions, you will need to define your own CDT _Build Variable_ to specify the GNU tool prefix for the timebeing:

Name:    `gnu_tool_prefix`
Type:      `String`
Value:     `riscv32-unknown-elf-`

Modify the value to suit your toolchain and be sure to include the trailing hyphen.

**Comment by baoruiqi** [comment]
I have previously configured the toolchain path on the property page of the project.

And I' sure no risc-v related toolchain is included in my PATH environment variable, so if eclipse tries to find the executable riscv64-unknown-elf-gcc from the PATH, I will get a not found prompt, but the fact is that the project can be correct build.

![image](https://github.com/eclipse-cdt/cdt/assets/128341204/49603a1a-f5e1-47e9-b8a8-80fd6b5b4465)

**Comment by baoruiqi** [comment]
> @baoruiqi, the `architecture: UNKNOWN!` output suggests that you are invoking a native `objdump` that exists somewhere on your PATH.
> 
> When using the _Embedded CDT_ toolchain descriptions, you will need to define your own CDT _Build Variable_ within the project properties dialog to specify the GNU tool prefix for the timebeing:
> 
> Name: `gnu_tool_prefix` Type: `String` Value: `riscv64-unknown-elf-`
> 
> Modify the value to suit your toolchain and be sure to include the trailing hyphen.

I have previously configured the toolchain path on the property page of the project.

And I' sure no risc-v related toolchain is included in my PATH environment variable, so if eclipse tries to find the executable riscv64-unknown-elf-gcc from the PATH, I will get a not found prompt, but the fact is that the project can be correct build.

![image](https://github.com/eclipse-cdt/cdt/assets/128341204/49603a1a-f5e1-47e9-b8a8-80fd6b5b4465)

**Comment by baoruiqi** [comment]
@jld01  I found the reason for the unknown architecture. When I first created the c managed project, I chose RISC-V cross gcc, which is provided by the embedded cdt project. And the cross toolchain provided by CDT is Cross gcc.

However, when I create a new project based on the cross gcc toolchain and set the binary parser to gnu elf parser, the result of double-clicking the elf file is a mess characters.

![image](https://github.com/eclipse-cdt/cdt/assets/128341204/d06841df-f9e7-446c-9c11-63e5aa3c8cba)

![image](https://github.com/eclipse-cdt/cdt/assets/128341204/8c3c6dbc-a209-41e1-b953-55a600170885)

![image](https://github.com/eclipse-cdt/cdt/assets/128341204/57998e55-a12f-41e6-8a08-3106519ed75f)

![image](https://github.com/eclipse-cdt/cdt/assets/128341204/eaf043e6-cb3c-430a-98be-3f02f885b459)

**Comment by jld01** [comment]
@baoruiqi does the _Default Binary File Editor_ open correctly if you select it explicitly from the context menu of your generated binary in the _Project Explorer_ view?

_Open With_ > _Other..._ > _Internal Editors_ > _Default Binary File Editor_

**Comment by jld01** [comment]
@baoruiqi I already have a patch for _Embedded CDT_ to generate the `gnu_tool_prefix` build variable. I will clean it up and create a PR.

**Comment by baoruiqi** [comment]
@jld01  Great, you are so familiar with Eclipse CDT, could you please give me some advice about this Issue?

https://github.com/eclipse-embed-cdt/eclipse-plugins/discussions/564

**Comment by jld01** [comment]
> ... when creating a cross gcc project, the generated binary file has no suffix by default. If I add the elf suffix to the current Build artifact on the project properties page, then it works.

@baoruiqi, this issue has been fixed via #409
